src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -3492,10 +3534,20 @@ private string ScanNumberHelper(char firstChar, out bool hex, out bool real, out
             if (c.IsTypeSuffix())
             {
                 SkipChar();
-                suffix = c;
+                suffix += c;
+                c = PeekChar();
+            }
+            if (c.IsTypeSuffix())
+            {
+                SkipChar();
+                suffix += c;
                 c = PeekChar();
             }
 
+            // Due to the suffix being a string instead of char, it must be null-terminated for console
+            // to behave correctly. If someone knows why, I will be happy to add a proper explanation.
+            suffix += "\0";
```

Comment:
This is horrible for perf - it will allocate a new string for every single number parsed.

It's probably cleaner to fully parse the suffix into flags here (as a flags enum) and pass that instead of a string.

Reply:
That is likely to be correct.

I guarantee you I do not have the knowledge at this point in time to be able to do that without some significant pointers from you fellas on how you would want that to be implemented.

Char based flag enum? Something like that?  I am game to give it a shot, but I've never worked with flag enums internally before. Need at the very least a good idea of the desired method of implementation to start on something like this.

Reply:
I was thinking of something like:

```c#
[Flags]
enum NumberSuffix
{
    None = 0x0,
    Long = 0x1,
    Double = 0x2,
    Unsigned = 0x4,
}
```

And by parsing, maybe something like:

```c#
            var suffix = NumberSuffix.None;
            c = PeekChar();
            if (c.IsTypeSuffix())
            {
                SkipChar();
                switch c {
                case 'l': case 'L':
                    suffix |= NumberSuffix.Long;
                    break;
                case 'd': case 'D':
                    suffix |= NumberSuffix.Double;
                    break;
                case 'u': case 'U':
                    suffix |= NumberSuffix.Unsigned;
                    break;
                }
                c = PeekChar();
                if (c.IsTypeSuffix())
                {
                    SkipChar();
                    switch c {
                    case 'l': case 'L':
                        suffix |= NumberSuffix.Long;
                        break;
                    case 'u': case 'U':
                        suffix |= NumberSuffix.Unsigned;
                        break;
                    }
                    c = PeekChar();
                }
            }
```

Reply:
Woah, that *is* neat! Thank you! I'll see how that goes!

Would it be most appropriate to have the enum in its own file, or as part of the `tokenizer.cs` file?

Reply:
I'm a fan of keeping related types in the same file, especially when they are small and not used elsewhere, so I'd keep it in `tokenizer.cs`.

Reply:
And note my sample code allows something it probably shouldn't - I'll let you figure out what that is and avoid the problem :)

Reply:
I can see it would allow `dl` or `du` suffixes. In fact, any suffix ending in `u` shouldn't be necessary or permitted.

It's also got `double` instead of `decimal` ;)

Reply:
Of course - `d` is `double` in C, C++, and C#, but not PowerShell.

Reply:
Yeah. Decimal is `m` in those, if I remember my C# correctly. PS just infers double from a non-integer real and only bothers specifying the rest, heh.

Reply:
OK think this is done. Building to check now...

Reply:
Oooh, fun, it was still parsing cases where I had `dl` and `ll` with... mixed results. Let's... *not* do that. 

Unless we *really* want `ll` -- it could be `long long` aka `biginteger` ;)

Reply:
I do like what @iSazonov suggested in using `Span` - it might even end up being less code, it's probably worth a look.

Reply:
After talking with some of the folks on the PS discord (who also seem to agree) I may give this a shot.

For clarity: are we talking about putting the entire `_script` into a span instead of string, and then rather than building substrings as happens in the number token helper method, I presume I'd have it pass around smaller pieces of that grabbed by index?

Reply:
Just the suffix would go in a `Span`.

Reply:
Huh. Seems like that would be... minimal in terms of gain, but I'll take a look.

Reply:
@lzybkr had a bit of a fiddle with it, but from what I can see, reworking the suffix as a two-char array is going to make significantly more fragile code. I'd rather not create a mess someone else has to deal with.

As it is, the use of switch statements and flags effectively make it relatively trivial to add an additional type suffix in the event we ever need any others included. Swapping this particular bit out for a `Span<char>`  isn't worth creating the fragile `if (suffix[0] == 'x' && suffix[1] != 0)` bits and pieces that would subsequently litter `TryGetNumberValue`

Reply:
The code is cleaner and more flexible using flags here, I think, unless there's a better way to check the values in the Span<char> that I'm missing.